File: src/contests/cg/puzzle/rust/mirrors.rs
// 🎮 CodinGame Puzzle - mirrors
// https://www.codingame.com/training/easy/mirrors

fn rl() -> String {
    let mut s = String::new();
    std::io::stdin().read_line(&mut s).unwrap();
    s.trim().to_string()
}

fn main() {
    let _ = rl();
    let r: Vec<f64> = rl().split_whitespace().map(|x| x.parse().unwrap()).collect();
    let mut reflected = 0.0f64;
    for i in (0..r.len()).rev() {
        let ri = r[i];
        let denom = 1.0 - ri * reflected;
        reflected = ri
            + if denom == 0.0 {
                0.0
            } else {
                ((1.0 - ri) * (1.0 - ri) * reflected) / denom
            };
    }
    println!("{:.4}", reflected);
}

File: src/contests/cg/puzzle/dart/mirrors.dart
// 🎮 CodinGame Puzzle - mirrors
// https://www.codingame.com/training/easy/mirrors

import 'dart:io';

void main() {
  stdin.readLineSync();
  List<double> r =
      stdin.readLineSync()!.trim().split(' ').map(double.parse).toList();
  double reflected = 0.0;
  for (int i = r.length - 1; i >= 0; i--) {
    double ri = r[i];
    double denom = 1 - ri * reflected;
    reflected =
        ri + (denom == 0 ? 0.0 : ((1 - ri) * (1 - ri) * reflected) / denom);
  }
  print(reflected.toStringAsFixed(4));
}
